ui/src/lib.rs

```diff
@@ -0,0 +1,847 @@
+        add_new_line(opcode_list.len());
+        let paragraph =
+            Paragraph::new(text_output).block(block_source_code).wrap(Wrap { trim: true });
+        f.render_widget(paragraph, area);
+    }
+
+    /// Draw the stack into the stack pane
+    fn draw_stack<B: Backend>(
+        f: &mut Frame<B>,
+        debug_steps: &[DebugStep],
+        current_step: usize,
+        area: Rect,
+    ) {
+        let stack_space =
+            Block::default().title(format!(" Stack: {} ", current_step)).borders(Borders::ALL);
+        let stack = &debug_steps[current_step].stack;
+        let min_len = usize::max(format!("{}", stack.len()).len(), 2);
+
+        let text: Vec<Spans> = stack
+            .iter()
+            .enumerate()
+            .map(|(i, stack_item)| {
+                Spans::from(Span::styled(
+                    format!("{: <min_len$}: {:?} \n", i, stack_item, min_len = min_len),
+                    Style::default().fg(Color::White),
+                ))
+            })
+            .collect();
+        let paragraph = Paragraph::new(text).block(stack_space).wrap(Wrap { trim: true });
+        f.render_widget(paragraph, area);
+    }
+
+    /// Draw memory in memory pane
+    fn draw_memory<B: Backend>(
+        f: &mut Frame<B>,
+        debug_steps: &[DebugStep],
+        current_step: usize,
+        area: Rect,
+    ) {
+        let memory = &debug_steps[current_step].memory;
+        let stack_space = Block::default()
+            .title(format!(" Memory - Max Expansion: {} bytes", memory.effective_len()))
+            .borders(Borders::ALL);
+        let memory = memory.data();
+        let max_i = memory.len() / 32;
+        let min_len = format!("{:x}", max_i * 32).len();
+
+        let text: Vec<Spans> = memory
+            .chunks(32)
+            .enumerate()
+            .map(|(i, mem_word)| {
+                let strings: String = mem_word
+                    .chunks(4)
+                    .map(|bytes4| {
+                        bytes4
+                            .iter()
+                            .map(|byte| {
+                                let v: Vec<u8> = vec![*byte];
+                                hex::encode(&v[..])
+                            })
+                            .collect::<Vec<String>>()
+                            .join(" ")
+                    })
+                    .collect::<Vec<String>>()
+                    .join("  ");
+                Spans::from(Span::styled(
+                    format!("{:0min_len$x}: {} \n", i * 32, strings, min_len = min_len),
+                    Style::default().fg(Color::White),
+                ))
+            })
+            .collect();
+        let paragraph = Paragraph::new(text).block(stack_space).wrap(Wrap { trim: true });
+        f.render_widget(paragraph, area);
+    }
+}
+
+impl Ui for Tui {
+    fn start(mut self) -> Result<TUIExitReason> {
+        let tick_rate = Duration::from_millis(75);
```

Comment:
Maybe this should be CLI configurable, in case somebody has a slower computer?

Reply:
ye i need to look up recommended timings from online